### ISSUE TITLE
additional terms

### ISSUE THREAD

**Issue by moimikey** [hemanth/functional-programming-jargon#20]
Hi,

I'd be more than happy to submit some pull requests on these various terms. I'd much rather get an opinion first before starting on a PR:

- [ ] application (as a compliment to `partial application`)
- [ ] thunk
- [ ] tail-call optimization
- [ ] trampolines
- [ ] functor
- [ ] reducer
- [ ] transducer
- [ ] homomorphism

**Comment by hsavit1** [comment]
@moimikey -  this would be a fantastic if you could define these!

I would add natural transformations and morphisms as well to this list

**Comment by moimikey** [comment]
awesome. will do :+1:

**Comment by moimikey** [comment]
@hsavit1 there's that fine line between functional programming paradigms and concepts that we use daily, vs. all and everything under category theory. this list could grow or get chopped down. i'll obviously crop and add and crop some more:
- [ ] morphisms
  - [ ] homo-
  - [ ] iso-
  - [ ] mono-
  - [ ] epi-
  - [ ] bi-
  - [ ] endo-
  - [ ] auto-
  - [ ] natural transformation

**Comment by hsavit1** [comment]
@moimikey agreed, we wouldn't want to "scare" anyone away from functional programming with overwhelming dictionary of category theory terms (and no examples)

on the other hand, I'd love to educate more programmers on category theory concepts. They were insanely helpful for me when learning Haskell for the first time.

**Comment by hemanth** [comment]
@moimikey Sorry for the late reply, please feel free to do the PRs! 

Thank you!

**Comment by jethrolarson** [comment]
Functor is already in. 

This is turning into a dictionary fast but maybe that's still okay. 
Some of the topics kind of deserve their own pages. Especially when it comes to defining why they're useful. I could see a page dedicated to morphisms.

It's hard to summarize monad in a paragraph and one or two examples don't illuminate, for example.

**Comment by hemanth** [comment]
@jethrolarson We could start a wiki in this repo, maybe?

**Comment by hemanth** [comment]
@jethrolarson Interested?

**Comment by flip111** [comment]
polymorphism

**Comment by stereobooster** [comment]
immutability and [functional data structures](https://www.cs.cmu.edu/~rwh/theses/okasaki.pdf)

**Comment by hemanth** [comment]
profunctors as well.

**Comment by stereobooster** [comment]
totality e.g. no effects at all. Exceptions and divergence as effects, but not side effects, e.g. "pure" function can have it. And side effects is IO and non-determinism.

**Comment by stereobooster** [comment]
`scopes` as soon as programing language has **symbols** it has to deal with scopes. Basically it can ignore them and make one global scope. Or choose functions as scopes and visibility of higher-call stack scopes (hello **closures**). Or lexical scopes, or module scopes. It is also to choose different models of scopes for different types of symbols.

Language without symbols - simple lambda calculus.
One global scope - lambda calculus extended with symbols or assembly language.

**Comment by tfausak** [comment]
Thanks for the great resource! Is this where I can report bugs on [this post](https://functional.works-hub.com/blog/Functional-Programming-Jargon)? If so:

- The titles don't define anchors (like `<h2 name="...">`), so the links in the table of contents don't go anywhere. 
- One of the paragraphs in the "Closure" section isn't formatted properly.

![capture](https://user-images.githubusercontent.com/408753/32497402-0ca56c98-c392-11e7-95c6-c412b5c21b19.PNG)

**Comment by jethrolarson** [comment]
functional.works-hub.com is not maintained by this repo. They copied the readme and didn't compile the markdown properly. I'm not sure who you should bug.

**Comment by nikelborm** [comment]
I think there are a bunch of new terms that could be added

- https://middle-ages.github.io/effect-ts-laws-docs/catalog-of-laws.html
- profunctor optics https://github.com/hablapps/DontFearTheProfunctorOptics/
- Affines, HKTs, Typeclasses
- https://oleg.fi/gists/posts/2017-04-18-glassery.html
  <img width="610" height="362" alt="Image" src="https://github.com/user-attachments/assets/a6605bf9-f664-4681-9207-8143fcf9701f" />
